cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/rolling/Batching.java

```diff
@@ -0,0 +1,302 @@
+                        availBatches.add(currentBatch);
+                    }
+                    currentBatch.add(kafkaNode);
+                } else {
+                    LOGGER.debugCr(reconciliation, "Some replicas of node {} will be unavailable => do not add to batch", kafkaNode.id());
+                    unavail.add(kafkaNode);
+                }
+            }
+            result.addAll(availBatches);
+        }
+        if (result.isEmpty() && !unavail.isEmpty()) {
+            LOGGER.warnCr(reconciliation, "Cannot restart nodes {} without violating some topics' min.in.sync.replicas", idsOf(unavail));
+        }
+        return result;
+    }
+
+    static <T> T elementInIntersection(Set<T> set, Set<T> set2) {
+        for (T t : set) {
+            if (set2.contains(t)) {
+                return t;
+            }
+        }
+        return null;
+    }
+
+    static boolean containsAny(Reconciliation reconciliation,
+                               KafkaNode node,
+                               Set<Replica> nodeReplicas,
+                               Set<KafkaNode> cell) {
+        for (var b : cell) {
+            var commonReplica = elementInIntersection(b.replicas(), nodeReplicas);
+            if (commonReplica != null) {
+                LOGGER.debugCr(reconciliation, "Nodes {} and {} have at least {} in common",
+                        node.id(), b.id(), commonReplica);
+                return true;
+            }
+        }
+        LOGGER.debugCr(reconciliation, "Node {} has no replicas in common with any of the nodes in {}",
+                node.id(), idsOf(cell));
+        return false;
+    }
+
+    private static String idsOf(Collection<KafkaNode> cell) {
+        return cell.stream()
+                .map(kafkaNode -> Integer.toString(kafkaNode.id()))
+                .collect(Collectors.joining(",", "{", "}"));
+    }
+
+    /** Returns a new set that is the union of each of the sets in the given {@code merge}. I.e. flatten without duplicates. */
+    private static <T> Set<T> union(Set<Set<T>> merge) {
+        HashSet<T> result = new HashSet<>();
+        for (var x : merge) {
+            result.addAll(x);
+        }
+        return result;
+    }
+
+
+
+    private static String idsOf2(Collection<? extends Collection<KafkaNode>> merge) {
+        return merge.stream()
+                .map(Batching::idsOf)
+                .collect(Collectors.joining(",", "{", "}"));
+    }
+
+    /**
+     * Pick the "best" batch to be restarted.
+     * This is the largest batch of available servers
+     * @return the "best" batch to be restarted
+     */
+    static Set<KafkaNode> pickBestBatchForRestart(List<Set<KafkaNode>> batches) {
+        var sorted = batches.stream().sorted(Comparator.comparing(Set::size)).toList();
+        if (sorted.size() == 0) {
+            return Set.of();
+        }
+        return sorted.get(sorted.size() - 1);
+    }
+
+
+    protected static boolean affectsAvailability(KafkaNode kafkaNode,
```

Comment:
Did we prefer the `anyReplicaWouldBeUnderReplicated`? It's a bit long, but it is more descriptive.

Reply:
In the old roller the name for this method is `wouldAffectAvailability`

Reply:
I think I favour the name from the current roller, `wouldAffectAvailability`.